exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,404 @@
+    char *errmsg = NULL;
+    int r = sqlite3_exec(db->ptr, sql, callback, rows, &errmsg);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+    }
+
+    push(exec->stack, rows);
+    free(sql);
+}
+
+void sqlite_close(TExecutor *exec)
+{
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        return;
+    }
+
+    int retval = sqlite3_close(db->ptr);
+    if (retval == SQLITE_OK) {
+        db->ptr = NULL;
+    } else if (retval == SQLITE_BUSY) {
+        exec->rtl_raise(exec, "SqlException", "Unfinalized prepared statements or unfinished backup objects exist.", BID_ZERO);
+    }
+}
+
+void sqlite_cursorDeclare(TExecutor *exec)
+{
+    char *query = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    TString *name = string_fromString(top(exec->stack)->string); pop(exec->stack);
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        free(query);
+        string_freeString(name);
+        return;
+    }
+
+    // It is important to note that query will be freed in freeCursor(), so we simply pass it along and don't bother to make a copy of it.
+    Cell *c = cell_createOtherCell(declareCursor(exec, db->ptr, query));
+
+    // It is again, important to note that name is also preserved, as it becomes the active KEY for the dictionary entry, so once again, we pass it along.
+    dictionary_addDictionaryEntry(Cursors, name, c);
+
+    push(exec->stack, cell_fromString(name));
+}
+
+void sqlite_cursorOpen(TExecutor *exec)
+{
+    Cell *name = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    Cell *c = dictionary_findDictionaryEntry(Cursors, name->string);
+    if (c == NULL) {
+        exec->rtl_raise(exec, "SqlException", TCSTR(name->string), BID_ZERO);
+        goto done;
+    }
+    openCursor(c->other);
+
+done:
+    cell_freeCell(name);
+}
+
+void sqlite_cursorFetch(TExecutor *exec)
+{
+    Cell *name = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    Cell *c = dictionary_findDictionaryEntry(Cursors, name->string);
+    if (c == NULL) {
+        exec->rtl_raise(exec, "SqlException", TCSTR(name->string), BID_ZERO);
+        goto done;
+    }
+    Cell *result = cell_createArrayCell(0);
+
+    push(exec->stack, cell_fromBoolean(fetchCursor(c->other, result)));
+    push(exec->stack, result);
+
+done:
+    cell_freeCell(name);
+}
+
+void sqlite_cursorClose(TExecutor *exec)
+{
+    Cell *name = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    Cell *c = dictionary_findDictionaryEntry(Cursors, name->string);
+    if (c == NULL) {
+        exec->rtl_raise(exec, "SqlException", TCSTR(name->string), BID_ZERO);
+        cell_freeCell(name);
+        return;
```

Comment:
Would you like to use a `goto` label here to avoid duplicating the cleanup code?

Reply:
Added a goto here.

Reply:
Did you? I don't see one.